bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+	pub enum SendError {
+		NotApplicable,
+		NotRoutable,
+		Transport,
+		DestinationUnsupported,
+		ExceedsMaxMessageSize,
+		MissingArgument,
+		Fees,
+	}
+
+	impl<T: Config> From<XcmpSendError> for Error<T> {
+		fn from(e: XcmpSendError) -> Self {
+			match e {
+				XcmpSendError::NotApplicable => Error::<T>::Send(SendError::NotApplicable),
+				XcmpSendError::Unroutable => Error::<T>::Send(SendError::NotRoutable),
+				XcmpSendError::Transport(_) => Error::<T>::Send(SendError::Transport),
+				XcmpSendError::DestinationUnsupported =>
+					Error::<T>::Send(SendError::DestinationUnsupported),
+				XcmpSendError::ExceedsMaxMessageSize =>
+					Error::<T>::Send(SendError::ExceedsMaxMessageSize),
+				XcmpSendError::MissingArgument => Error::<T>::Send(SendError::MissingArgument),
+				XcmpSendError::Fees => Error::<T>::Send(SendError::Fees),
+			}
+		}
+	}
+
+	/// The nonce of the message been processed or not
+	#[pallet::storage]
+	pub type NonceBitmap<T: Config> = StorageMap<_, Twox64Concat, u128, u128, ValueQuery>;
+
+	/// The current operating mode of the pallet.
+	#[pallet::storage]
+	pub type OperatingMode<T: Config> = StorageValue<_, BasicOperatingMode, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Submit an inbound message originating from the Gateway contract on Ethereum
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::submit())]
+		pub fn submit(origin: OriginFor<T>, message: Message) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			ensure!(!OperatingMode::<T>::get().is_halted(), Error::<T>::Halted);
+
+			// submit message to verifier for verification
+			T::Verifier::verify(&message.event_log, &message.proof)
+				.map_err(|e| Error::<T>::Verification(e))?;
+
+			// Decode event log into an Envelope
+			let envelope =
+				Envelope::try_from(&message.event_log).map_err(|_| Error::<T>::InvalidEnvelope)?;
+
+			// Verify that the message was submitted from the known Gateway contract
+			ensure!(T::GatewayAddress::get() == envelope.gateway, Error::<T>::InvalidGateway);
+
+			// Verify the message has not been processed
+			ensure!(!Nonce::<T>::get(envelope.nonce.into()), Error::<T>::InvalidNonce);
+
+			let origin_account_location = Self::account_to_location(who)?;
+
+			let (xcm, _relayer_reward) =
+				Self::do_convert(envelope.message, origin_account_location.clone())?;
+
+			// Todo: Deposit fee(in Ether) to RewardLeger which should cover all of:
+			// T::RewardLeger::deposit(who, relayer_reward.into())?;
+			// a. The submit extrinsic cost on BH
+			// b. The delivery cost to AH
+			// c. The execution cost on AH
+			// d. The execution cost on destination chain(if any)
+			// e. The reward
+
+			// Set nonce flag to true
+			log::info!(target: "snowbridge-inbound-queue:v2","💫 setting nonce to {:?}", envelope.nonce);
+			Nonce::<T>::set(envelope.nonce.into());
+
+			// Attempt to forward XCM to AH
+			let message_id = Self::send_xcm(xcm, T::AssetHubParaId::get())?;
```

Comment:
any error will in practice revert/rollback storage changes, but let's be paranoid and mark nonce as processed only on certainty of success:

```suggestion
			// Attempt to forward XCM to AH
			let message_id = Self::send_xcm(xcm, T::AssetHubParaId::get())?;

			// Set nonce flag to true
			log::info!(target: "snowbridge-inbound-queue:v2","💫 setting nonce to {:?}", envelope.nonce);
			Nonce::<T>::set(envelope.nonce.into());
```

Reply:
fixed in https://github.com/paritytech/polkadot-sdk/pull/7402/commits/6f769ccdfd222654a26ce155f40468681d5ded28

Reply:
@yrong mentioned if we increment the nonce after the send we risk re-entrancy attacks.

Reply:
@vgeddes this is still an issue IMO, nonce should be incremented before the XCM is sent to prevent re-entrancy attacks: https://github.com/vgeddes/polkadot-sdk/blob/snowbridge-v2/bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs#L267